src/motors.c

```diff
@@ -64,91 +293,176 @@ extern void motor_set_speed_percent(motors_t motor_x, uint8_t speed, direction_t
-{
-	return 0;
+
+/*
+ * Returns motor RPM
+ * Param: Motor number
+ * Returns: Motor RPM, signed 16-bit int, positive value = forward rotation, negative value = reverse rotation
+ *          Max RPM value = 32767. If actual rpm exceeds this, this value will overflow.
+ */
+extern int16_t motor_get_rpm(motors_t motor_x) {
+	/*
+	 * Below is the motor to Timer mapping for RPM
+	 * Motor 1 - TIM5 CH2
+	 * Motor 2 - TIM5 CH3
+	 * Motor 3 - TIM1 CH4
+	 * Motor 4 - TIM1 CH2
+	 * Motor 5 - TIM4 CH3
+	 * Motor 6 - TIM4 CH2
+	 * Motor 7 - TIM3 CH2
+	 * Motor 8 - TIM3 CH4
+	 */
+	static uint16_t *motorToTIM_DIER[NUMBER_OF_MOTORS] = { &TIM5->DIER,
+														   &TIM5->DIER,
+														   &TIM1->DIER,
+														   &TIM1->DIER,
+														   &TIM4->DIER,
+														   &TIM4->DIER,
+														   &TIM3->DIER,
+														   &TIM3->DIER };
+
+	static uint16_t motorToTIM_CCxDE[NUMBER_OF_MOTORS] = { TIM_IT_CC2DE,
+														   TIM_IT_CC3DE,
+														   TIM_IT_CC4DE,
+														   TIM_IT_CC2DE,
+														   TIM_IT_CC3DE,
+														   TIM_IT_CC2DE,
+														   TIM_IT_CC2DE,
+														   TIM_IT_CC4DE };
+
+	/*
+	 * Below is the motor to DMA mapping for RPM
+	 * Motor 1 - DMA1 Stream4 CH6
+	 * Motor 2 - DMA1 Stream0 CH6
+	 * Motor 3 - DMA2 Stream4 CH6
+	 * Motor 4 - DMA2 Stream2 CH6
+	 * Motor 5 - DMA1 Stream7 CH2
+	 * Motor 6 - DMA1 Stream3 CH2
+	 * Motor 7 - DMA1 Stream5 CH5
+	 * Motor 8 - DMA1 Stream2 CH5
+	 */
+	static uint32_t *motorToDMA_NTDR[NUMBER_OF_MOTORS] = { &DMA1_Stream4->NDTR,
+														   &DMA1_Stream0->NDTR,
+														   &DMA2_Stream4->NDTR,
+														   &DMA2_Stream2->NDTR,
+														   &DMA1_Stream7->NDTR,
+														   &DMA1_Stream3->NDTR,
+														   &DMA1_Stream5->NDTR,
+														   &DMA1_Stream2->NDTR };
+
+	//need to stop raw data from updating
+	*motorToTIM_DIER[motor_x - 1] &= ~motorToTIM_CCxDE[motor_x - 1];
+
+	uint8_t idx = PWM_IN_ARRAY_LENGTH - (*motorToDMA_NTDR[motor_x - 1]); //Index of array element that was most recently updated
+	uint8_t stop_idx = (idx + PWM_IN_ARRAY_LENGTH - 1) % PWM_IN_ARRAY_LENGTH; //Index of array element that is most out of date
+	float frequency = 0;
+	int16_t rpm = 0;
+
+	for (; idx != stop_idx; idx = ((idx + 1) % PWM_IN_ARRAY_LENGTH)) {
+		uint32_t diff = 0; // Temporarily store the difference between two timestamps
+
+		if (pwmInTimestamp[motor_x - 1][(idx + 1) % PWM_IN_ARRAY_LENGTH] > pwmInTimestamp[motor_x - 1][idx]) {
+			diff = pwmInTimestamp[motor_x - 1][(idx + 1) % PWM_IN_ARRAY_LENGTH]	- pwmInTimestamp[motor_x - 1][idx];
+
+		} else { //if not bigger, must be smaller
+			diff = (_100_MHZ - pwmInTimestamp[motor_x - 1][idx]) + pwmInTimestamp[motor_x - 1][(idx + 1) % PWM_IN_ARRAY_LENGTH];
+		}
+		frequency += diff; // Add the differences. Will be averaged later
+	}
+
+	//make sure to turn the DMA back on
+	*motorToTIM_DIER[motor_x - 1] |= motorToTIM_CCxDE[motor_x - 1];
```

Comment:
When does DMA get turned off? Why does it need to be turned off?

Reply:
DMA gets turned off in line 355. The timer register that enables the DMA requests gets turned off. I turn it off so that the timestamps in the array don't update while I'm reading from it and processing it.